### PR TITLE
Video endpoint

### PR DESCRIPTION
**Pull request by VeronikaSolovei9**
-Added stubbed endpoint: Http method: POST, url mapping "/openrtb2/video"

### CODE REVIEW

File: endpoints/openrtb2/video_auction.go [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+	if err != nil {
+		return
+	}
+	storedRequest, err := loadStoredVideoRequest(storedRequestId)
+	if err != nil {
+		return
+	}
+
+	//merge incoming req with stored video req
+	resolvedRequest, err := jsonpatch.MergePatch(requestJson, storedRequest)
```
Comment (jmaynardxandr) [inline endpoints/openrtb2/video_auction.go]: I think the parameters should be swapped in this call, we want to make sure the requestJson overrides the storedRequest.

File: endpoints/openrtb2/video_auction.go [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+	//unmarshal and validate combined result
+	videoBidReq, errl := deps.parseVideoRequest(resolvedRequest)
+	if len(errl) > 0 {
+		return
+	}
+
+	var bidReq = openrtb.BidRequest{}
+
+	// Populate any "missing" OpenRTB fields with info from other sources, (e.g. HTTP request headers).
+	deps.setFieldsImplicitly(r, &bidReq)
```
Comment (jmaynardxandr) [inline endpoints/openrtb2/video_auction.go]: Hans was pointing out that we should swap the order of steps 6 and 7 in that write-up, so we should call setFieldsImplicitly after the mergeData step. That function will only overwrite fields that haven't already been set, so it could result in less overall work if we call it after the merge.

File: endpoints/openrtb2/video_auction.go [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+		0,
+		nil,
+		0,
+		nil,
+		nil,
+		nil,
+		nil,
+		nil,
+		nil,
+	}
```
Comment (jmaynardxandr) [inline endpoints/openrtb2/video_auction.go]: You should be able to omit those parameters, undeclared fields get assigned their zero value by default in Go.

File: endpoints/openrtb2/video_auction.go (hunk 1 above)
Comment (VeronikaSolovei9) [reply endpoints/openrtb2/video_auction.go]: Done

File: endpoints/openrtb2/video_auction.go (hunk 2 above)
Comment (VeronikaSolovei9) [reply endpoints/openrtb2/video_auction.go]: Done

File: endpoints/openrtb2/video_auction.go (hunk 3 above)
Comment (VeronikaSolovei9) [reply endpoints/openrtb2/video_auction.go]: Done

**Comment by hhhjort** [comment]
test

File: endpoints/openrtb2/video_auction.go [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+	requestJson, err := ioutil.ReadAll(lr)
+	if err != nil {
+		errL := []error{err}
+		handleError(labels, w, errL, ao)
+		return
+	}
+
+	//load additional data - stored simplified req
+	storedRequestId, err := getVideoStoredRequestId(requestJson)
```
Comment (jmaynardxandr) [inline endpoints/openrtb2/video_auction.go:102]: Can you add a TODO comment here so we remember that we are going to be adding a config option to say whether or not a storedRequestId is required? We can do the implementation for that in a different PR but we wanted to eventually make it configurable.

File: openrtb_ext/bid_request_video.go [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+	//   Site where the impression will be shown
+	Site openrtb.Site `json:"site"`
+
+	// Attribute:
+	//   user
+	// Type:
+	//   object; optional
+	// Description:
+	//   Container object for the user of of the actual device
+	User SimplifiedUser `json:"user,omitempty"`
```
Comment (jmaynardxandr) [inline openrtb_ext/bid_request_video.go:54]: The "omitempty" annotation should be used on pointers to structs rather than nested structs so value can be nil if not given when unmarshalling or completely omitted if nil when marshalling. Note this is the case for many fields in these structs, however "omitempty" is ok for a bool/int/string field that would default to false/zero/"" if it doesn't exist in the json.

File: openrtb_ext/bid_response_video.go [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+	Errors    VideoErrors      `json:"errors"`
+}
+
+type VideoTargeting struct {
+	Hb_pb         string `json:"hb_pb"`
+	Hb_pb_cat_dur string `json:"hb_pb_cat_dur"`
+	Hb_cache_id   string `json:"hb_cache_id"`
+}
+
+type VideoErrors struct {
```
Comment (jmaynardxandr) [inline openrtb_ext/bid_response_video.go:19]: We might want this to be a map[string]json.RawMessage since OpenX isn't the only bidder that may return errors. I'm not totally sure though if there is a consistent way it's done like if we can define a struct for the error messages that all bidders abide by or if we need to just allow it to be a generic json blob.

File: server/server.go [hunk 7]
```diff
@@ -72,12 +73,18 @@ func newAdminServer(cfg *config.Configuration, handler http.Handler) *http.Serve
 }
 
 func newMainServer(cfg *config.Configuration, handler http.Handler) *http.Server {
+	var serverHandler = handler
+	if cfg.EnableGzip {
+		serverHandler = gziphandler.GzipHandler(handler)
```
Comment (jmaynardxandr) [inline server/server.go:78]: Odd, where did these changes come from?

**Review by jmaynardxandr — APPROVED** [review]
This looks pretty good for the most part but I added comments in a couple of places, but if you can't get to it today we can merge into our fork's develop branch and address those comments at a later time. Thanks!
